File: src/solutions/day1.rs
use std::fmt::Display;
use crate::Solution;

pub struct Day1;

impl Day1 {
    pub fn part_one(&self, inp: impl Display) -> u32 {
        inp.to_string()
            .lines()
            .map(|line| {
                let digits = line.chars()
                    .into_iter()
                    .filter_map(|c| c.to_digit(10))
                    .collect::<Vec<u32>>();
                digits.first().unwrap() * 10 +
                digits.last().unwrap()
            })
            .sum()
    }

    pub fn part_two(&self, inp: impl Display) -> u32 {
        let mut inp = inp.to_string();
        let map = [
            ("one", "1"),
            ("two", "2"),
            ("three", "3"),
            ("four", "4"),
            ("five", "5"),
            ("six", "6"),
            ("seven", "7"),
            ("eight", "8"),
            ("nine", "9"),
        ];
        for (key, val) in map {
            inp = inp.replace(key, format!("{key}{val}{key}").as_str());
        }
        self.part_one(inp)
    }
}

impl Solution for Day1 {
    const NAME: &'static str = "Trebuchet!?";

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 53651);
        assert_eq!(p2, 53894);
    }
}

File: aoc-py/solutions/day1.py
"""
Day 1: Trebuchet?!

https://adventofcode.com/2023/day/1
"""
__all__ = ('Day1',)

from typing import ClassVar
from .. import run_day
from ..solution import Solution

class Day1(Solution):
    NAME = 'Trebuchet?!'
    NUM_MAP: ClassVar[dict[str, str]] = {
        'one': '1',
        'two': '2',
        'three': '3',
        'four': '4',
        'five': '5',
        'six': '6',
        'seven': '7',
        'eight': '8',
        'nine': '9'
    }

    def part_one(self, inp: str) -> int:
        return sum(
            int(
                (digits := [c for c in line if c.isdigit()])[0]
                + digits[-1]
            )
            for line in inp.splitlines()
        )

    def part_two(self, inp: str) -> int:
        for key in self.NUM_MAP:
            inp = inp.replace(key, key + self.NUM_MAP[key] + key)
            # accounts for overlapping words:
            #   i.e. 'twone'
            #   -> 'twone1one' (when key = 'one')
            #   -> '2ne1one'   (when key = 'two')
            #   -> 21 (correct)
            #
            # Without appending `key` to either side:
            #   'twone' -> '2ne' -> 2 (wrong)
        return self.part_one(inp)

    def run(self, inp: str) -> None:
        print('Part 1: ', p1 := self.part_one(inp))
        print('Part 2: ', p2 := self.part_two(inp))

        assert p1 == 53651
        assert p2 == 53894

if __name__ == '__main__':
    run_day(1)